tests/func/test_lock.py

```diff
@@ -52,3 +55,38 @@ def test_cli(tmp_dir, dvc, mocker, caplog):
     with Lock(tmp_dir / dvc.tmp_dir / "lock"):
         assert main(["add", "foo"]) == 1
     assert expected_error_msg in caplog.text
+
+
+def hold_lock(lockfile_path, hold_seconds):
+    lock = make_lock(lockfile_path)
+    with lock:
+        time.sleep(hold_seconds)
+
+
+def try_lock_with_wait(lockfile_path, wait, result_queue):
+    try:
+        lock = make_lock(lockfile_path, wait=wait)
+        with lock:
+            result_queue.put("acquired")
+    except LockError as e:
+        result_queue.put(f"error: {e}")
+
+
+def test_lock_waits_when_requested(tmp_path):
+    lockfile = tmp_path / "lock"
+
+    # Process 1 holds the lock for 2 seconds
+    p1 = multiprocessing.Process(target=hold_lock, args=(lockfile, 2))
+    p1.start()
+    time.sleep(0.2)  # let p1 acquire lock
+
+    # Process 2 will wait for the lock (should succeed)
+    q = multiprocessing.Queue()
+    p2 = multiprocessing.Process(target=try_lock_with_wait, args=(lockfile, True, q))
+    p2.start()
+    p2.join(timeout=4)
```

Comment:
Do we need multiprocessing here? Can we run this in the same process, by creating two instances of `Lock`?

Reply:
Cheers for the review!  On this point, unless there's some concurrency stuff happening under the hood in the library, I _think_ we might need some sort of concurrency.

The reason I did it is that I want to test that:

1. P2 will wait without crashing while P1 still holds the lock
2. P2 will succeed after P1 releases the lock

I reckon `make_lock` is blocking, so I don't think there's a way that I could test 1, and then release the lock to test 2, without some sort of multiprocessing.